File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    const q = [];
    const visited = Array.from({length : n},()=>new Array(m).fill(false));
    q.push([0,0,1])
    visited[0][0] = true;
    let idx = 0;
    const d = [[1,0], [-1,0], [0,-1], [0,1]];
    
    while(idx < q.length){
        const [cN,cM,pt] = q[idx++];
        if(cN === n - 1 && cM === m - 1) return pt;
        for(let i = 0; i < d.length; i++){
            const [dN,dM] = d[i];
            const nN = cN + dN;  
            const nM = cM + dM;
            if( nM >= m || nM < 0 || nN >= n || nN < 0 || visited[nN][nM] || maps[nN][nM] === 0) continue;
            q.push([nN,nM,pt+1]);
            visited[nN][nM] = true;
        }
    }
 return -1;   
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    q = deque()
    n = len(maps)
    m = len(maps[0])
    dr = [1, -1, 0, 0]
    dc = [0, 0, 1, -1]
    visited = [[False] * m for _ in range(n)]
    visited[0][0] = True
    q.append([0,0,1])
    
    while q:
        r, c, cnt = q.popleft()
        if r == n -1 and c == m-1:
            return cnt
        
        for d in range(4):    
            nxt_r, nxt_c = r + dr[d], c + dc[d]
            if nxt_r < 0 or nxt_r >= n or nxt_c < 0 or nxt_c >= m:
                continue
            if maps[nxt_r][nxt_c] == 0 or visited[nxt_r][nxt_c]:
                continue
            q.append([nxt_r, nxt_c, cnt + 1])
            visited[nxt_r][nxt_c] = True

    return -1